source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1112 @@
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to resolve role session name during sts web identity provider initialization.")
+            goto on_finish;
+        }
+    } else {
+        struct aws_uuid uuid;
+
+        if (aws_uuid_init(&uuid)) {
+            AWS_LOGF_ERROR(AWS_LS_IO_PKI, "Failed to load an uuid during sts web identity provider initialization.");
+            goto on_finish;
+        }
+
+        char uuid_str[AWS_UUID_STR_LEN] = {0};
+        struct aws_byte_buf uuid_buf = aws_byte_buf_from_array(uuid_str, sizeof(uuid_str));
+        uuid_buf.len = 0;
+        if (aws_uuid_to_str(&uuid, &uuid_buf)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_IO_PKI,
+                "Failed to stringify uuid for role session name during sts web identity provider initialization.");
+            goto on_finish;
+        }
+        if (aws_byte_buf_init_copy(&parameter->role_session_name, allocator, &uuid_buf)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to generate role session name during sts web identity provider initialization.")
+            goto on_finish;
+        }
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "Successfully loaded all required parameters for sts web identity credentials provider.")
+    success = true;
+
+on_finish:
+    if (region) {
+        aws_string_destroy(region);
+    }
+    if (role_arn) {
+        aws_string_destroy(role_arn);
+    }
+    if (role_session_name) {
+        aws_string_destroy(role_session_name);
+    }
+    if (token_file_path) {
+        aws_string_destroy(token_file_path);
+    }
+    if (profile_name) {
+        aws_string_destroy(profile_name);
+    }
+    if (config_profile) {
+        aws_profile_collection_destroy(config_profile);
+    }
+    if (!success) {
+        aws_mem_release(allocator, parameter);
+        parameter = NULL;
+    }
+    return parameter;
+}
+
+void s_sts_web_identity_parameters_destroy(struct aws_credentials_provider_sts_web_identity_parameters *parameter) {
+    if (!parameter) {
+        return;
+    }
+    aws_byte_buf_clean_up(&parameter->endpoint);
+    aws_byte_buf_clean_up(&parameter->role_arn);
+    aws_byte_buf_clean_up(&parameter->role_session_name);
+    aws_byte_buf_clean_up(&parameter->token_file_path);
+    aws_mem_release(parameter->allocator, parameter);
+}
+
+struct aws_credentials_provider *aws_credentials_provider_new_sts_web_identity(
+    struct aws_allocator *allocator,
+    const struct aws_credentials_provider_sts_web_identity_options *options) {
+
+    struct aws_credentials_provider_sts_web_identity_parameters *parameters =
+        s_sts_web_identity_parameters_new(allocator);
+    if (!parameters) {
+        return NULL;
+    }
+
+    struct aws_credentials_provider *provider = NULL;
+    struct aws_credentials_provider_sts_web_identity_impl *impl = NULL;
+
+    aws_mem_acquire_many(
```

Comment:
just FYI, a simpler way to put everything on the same allocation is to put the base class as a member by value on your implementation struct, then return a pointer to it.

Reply:
+1 to this

Reply:
Then the `impl` and `provider` base class will reference each other : one by pointer, another by value. It's weird though in impl's clean up, we won't clean up provider's memory, then in "extended" provider's clean up, we will clean up impls'. I think this trick is a bit confusing and error prone, one has to deal the mem alloc and dealloc very carefully.